backends/ubpf/ubpfRegister.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2019 Orange
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "ubpfRegister.h"
+
+namespace UBPF {
+
+    static cstring last_key_name;
+
+    UBPFRegister::UBPFRegister(const UBPFProgram *program,
+                               const IR::ExternBlock *block,
+                               cstring name, EBPF::CodeGenInspector *codeGen) :
+            UBPFTableBase(program, name, codeGen) {
+
+        auto di = block->node->to<IR::Declaration_Instance>();
+        auto type = di->type->to<IR::Type_Specialized>();
+        auto valueType = type->arguments->operator[](0);
+        auto keyType = type->arguments->operator[](1);
+
+        this->valueType = valueType;
+        this->keyType = keyType;
+
+        if (valueType->is<IR::Type_Name>()) {
```

Comment:
Probably the right way to do this is to look it up in the TypeMap.

Reply:
Could you elaborate more on that? I'm not sure I got the point.

Reply:
This should have been a bit higher: instead of using `di->type` you should do `typeMap->getType(di)`.

Reply:
Done.